subgraph-js/src/schema-helper/transformSchema.ts

```diff
@@ -117,9 +117,9 @@ export function transformSchema(
   function replaceType(type: GraphQLInputType): GraphQLInputType;
   function replaceType(type: GraphQLType): GraphQLType {
     if (isListType(type)) {
-      return new GraphQLList(replaceType(type.ofType));
+      return new GraphQLList(replaceType(type.ofType as any));
     } else if (isNonNullType(type)) {
-      return new GraphQLNonNull(replaceType(type.ofType));
+      return new GraphQLNonNull(replaceType(type.ofType as any));
```

Comment:
This was what I remembered running into. I'll take a closer look, I'm pretty curious about this.

Reply:
It's non ideal that `GraphQLType` is a union type where two of the possible types are `GraphQLList<any>` and `GraphQLNonNull<any>` but I'm not sure what to do about it exactly.

Reply:
Not `<any>`, they just recurse on themselves. I _think_ I've got a solution, here's a patch I came up with. The main diff is adding one more "overload" to the signature which is the _actual_ signature itself. In the recursive case (list / null) it didn't seem to recognize that any `GraphQLType` was acceptable to be passed in.

Aside from the overload, I only added a bit more info about the return types.

```diff
diff --git a/subgraph-js/src/schema-helper/transformSchema.ts b/subgraph-js/src/schema-helper/transformSchema.ts
index e9ea533d..b845cdc6 100644
--- a/subgraph-js/src/schema-helper/transformSchema.ts
+++ b/subgraph-js/src/schema-helper/transformSchema.ts
@@ -115,11 +115,12 @@ export function transformSchema(
   function replaceType(type: GraphQLNamedType): GraphQLNamedType;
   function replaceType(type: GraphQLOutputType): GraphQLOutputType;
   function replaceType(type: GraphQLInputType): GraphQLInputType;
+  function replaceType(type: GraphQLType): GraphQLType;
   function replaceType(type: GraphQLType): GraphQLType {
     if (isListType(type)) {
-      return new GraphQLList(replaceType(type.ofType as any));
+      return new GraphQLList(replaceType(type.ofType));
     } else if (isNonNullType(type)) {
-      return new GraphQLNonNull(replaceType(type.ofType as any));
+      return new GraphQLNonNull(replaceType(type.ofType));
     }
     return replaceNamedType(type);
   }
@@ -154,19 +155,23 @@ export function transformSchema(
     }));
   }
 
-  function replaceArgs(args: GraphQLFieldConfigArgumentMap) {
-    return mapValues(args, arg => ({
+  function replaceArgs(
+    args: GraphQLFieldConfigArgumentMap,
+  ): GraphQLFieldConfigArgumentMap {
+    return mapValues(args, (arg) => ({
       ...arg,
-      type: replaceType(arg.type)
+      type: replaceType(arg.type),
     }));
   }
 
-  function replaceDirectives(directives: readonly GraphQLDirective[]) {
-    return directives.map(directive => {
+  function replaceDirectives(
+    directives: readonly GraphQLDirective[],
+  ): readonly GraphQLDirective[] {
+    return directives.map((directive) => {
       const config = directive.toConfig();
       return new GraphQLDirective({
         ...config,
-        args: replaceArgs(config.args)
+        args: replaceArgs(config.args),
       });
     });
   }
```

Reply:
So this looks great, but I don't understand why it's working. I'll have to have you walk me through it tomorrow.